src/main/java/build/buildfarm/worker/CASFileCache.java

```diff
@@ -0,0 +1,238 @@
+    this.instance = instance;
+    this.root = root;
+    this.maxSizeInBytes = maxSizeInBytes;
+    paths = new HashMap<>();
+    accessScanPaths = new ArrayList<>();
+    sizeInBytes = 0;
+  }
+
+  public void start() throws IOException {
+    Files.createDirectories(root);
+
+    Files.walkFileTree(root, new SimpleFileVisitor<Path>() {
+      @Override
+      public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+        long size = attrs.size();
+        if (sizeInBytes + size > maxSizeInBytes) {
+          // FIXME need to do this according to expiration strategy
+          Files.delete(file);
+        } else {
+          paths.put(file, new FileEntry(
+              file,
+              size,
+              attrs.lastModifiedTime().to(TimeUnit.NANOSECONDS),
+              attrs.lastAccessTime().to(TimeUnit.NANOSECONDS)));
+          sizeInBytes += size;
+        }
+        return FileVisitResult.CONTINUE;
+      }
+    });
+    update();
+  }
+
+  public void update() {
+    for (Path path : accessScanPaths) {
+      FileEntry entry = paths.get(path);
+      if (entry != null) {
+        try {
+          BasicFileAttributes attrs = Files.readAttributes(path, PosixFileAttributes.class, NOFOLLOW_LINKS);
+          /* should verify invariants, are */
+          paths.get(path).setAccessTime(attrs.lastAccessTime().to(TimeUnit.NANOSECONDS));
+        } catch (IOException ex) {
+          // FIXME report error
+        }
+      }
+    }
+    accessScanPaths.clear();
+    expirationQueue = Queues.newPriorityQueue(paths.values());
+  }
+
+  private void expireEntry() {
+    FileEntry entry = expirationQueue.remove();
+
+    Path path = entry.getPath();
+    try {
+      Files.delete(path);
+      paths.remove(path);
+      sizeInBytes -= entry.getSizeInBytes();
+    } catch (IOException ex) {
+      // FIXME report error
+    }
+  }
+
+  public Path put(Digest digest, boolean isExecutable) {
+    long size = digest.getSizeBytes();
+    Path path = root.resolve(String.format(
+        "%s_%d%s", digest.getHash(), size, isExecutable ? "_exec" : ""));
+
+    if (paths.containsKey(path)) {
+      accessScanPaths.add(path);
+      return path;
+    }
+
+    // the cache cannot hold this file
+    if (size > maxSizeInBytes) {
+      throw new IllegalStateException();
+    }
+
+    // remove cache nodes until we have cache size + digest.size_bytes <= limit
+    while (sizeInBytes + size > maxSizeInBytes && !expirationQueue.isEmpty()) {
+      // FIXME could have no entries left to remove...
+      expireEntry();
+    }
+
+    if (sizeInBytes + size > maxSizeInBytes) {
+      System.out.println(String.format(
+          "Expiration did not yield sufficient space: %d + %d > %d",
+          sizeInBytes,
+          size,
+          maxSizeInBytes));
+      return null;
+    }
+
+    try (InputStream in = instance.newStreamInput(instance.getBlobName(digest))) {
```

Comment:
This seems like dependency inversion. I'd expect a class called CASFileCache to not have a reference back to the Instance, but to work more as a library class.

Reply:
Consider passing in an InputStream or some such. Also, if we're eventually switching to store files on disk in the instance, we might want to not have an intermediate copy here.

Reply:
I'll adjust to use an inputStreamFactory instead of the instance to clean up the dependencies.
While I know the origin of `eventually switching to store files on disk in the instance`, I want to head there in reasonable steps.  The only intermediate copy here may eventually be replaced by a Path response from the instance, but let's cross that bridge when we set fire to this one.